### PR TITLE
Introduce securityCodes for keyboard mode

### PR DESCRIPTION
**Pull request by gasiortomasz**

### CODE REVIEW

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 1]
```diff
@@ -32,6 +28,25 @@ export const isValidCode = (code: number[], publicKey: string) => {
   return deepEqual(expectedCode, code);
 };
 
+export const findValidAddressFromPartCode = (code: number[], incomingPublicKeys: string[]) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]: PartCode -> CodePart

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 2]
```diff
@@ -32,6 +28,25 @@ export const isValidCode = (code: number[], publicKey: string) => {
...
+  const possibleAddresses = findPossibleAddressesFromPartCode(code, incomingPublicKeys);
+  return possibleAddresses.length === 1 ? possibleAddresses[0] : undefined;
+};
+
+export const findPossibleAddressesFromPartCode = (code: number[], incomingPublicKeys: string[]) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]: PartCode -> CodePart, code -> codePart

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 3]
```diff
@@ -32,6 +28,25 @@ export const isValidCode = (code: number[], publicKey: string) => {
...
+  return incomingPublicKeys.filter((incomingPublicKey) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]:
```ts
arrayStartWith(prefix, array) =>
  if (prefix.lenght > array.length) {
    return false;
  } 
  for (i = 0...) {
    if (array[i] !== prefix[i]) {
      return false
    }
  }
  return true;
}
```
or
```ts
deepArrayStartWith(prefix, array) => {
  if (prefix.lenght > array.length) {
    return false;
  } 
 const arrayPrefix = array.slice(0, prefix.length);
 return deepEqual(arrayPrefix, prefix); 
}
```

File: universal-login-commons/lib/core/utils/securityCodes.ts (hunk 2 above)
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]:
```ts
findCodesByPrefixAndAddresses(prefix, address) => 
  findCodesByPrefix(codes.map((key) => generateCode(key)))

findCodesByPrefix(prefix, codes) => 
  codes.filter((code) => arrayStartsWith(prefix, code))
```

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 4]
```diff
@@ -18,11 +18,7 @@ export const generateCodeWithFakes = (publicKey: string): number[] => {
 };
 
 export const generateCode = (publicKey: string): number[] => {
-  const hash = utils.keccak256(publicKey);
-  const bytes8bit = Array.from(utils.arrayify(hash));
-  const bytes16bit = array8bitTo16bit(bytes8bit);
-
-  return bytes16bit
+  return addressTo16bitBytes(publicKey)
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]: 👍

File: universal-login-commons/test/core/utils/securityCodes.ts [hunk 5]
```diff
@@ -36,6 +36,70 @@ describe('UNIT: security codes', () => {
...
+    ];
+    const addressesThatShareFirstCodeWithRealAddress = [
+      '0x49c9A6784C061D298f9021a07eC218382feE20A9',
+      '0xf247e3c2f118763f79BE7C226D1c3dB988004704',
+      '0x22A6F13A956e7235fd989329da6e504523EE9346',
+      '0xd856550341C0772BC8f02E8E30DcD43239677084',
+      '0x69a4de9CF98FFf3D019F85C2d0aeb3860C179EBC'
+    ];
+
+    const attackerAddresses = addressesThatDontShareBytesWithRealAddress.concat(addressesThatShareFirstCodeWithRealAddress);
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/securityCodes.ts]: const attackerAddresses = [...addressesThatDontShareBytesWithRealAddress, ...addressesThatShareFirstCodeWithRealAddress];
const incomingAddresses = [...attackerAddresses, realAddress];

File: universal-login-commons/test/core/utils/securityCodes.ts [hunk 6]
```diff
@@ -36,6 +36,70 @@ describe('UNIT: security codes', () => {
     });
   });
 
+  describe('keyboard attack mode', () => {
+    const realAddress = '0xee2C70026a0E36ccC7B9446b57BA2bD98c28930b';
+    const expectedCode = generateCode(realAddress);
+    const addressesThatDontShareBytesWithRealAddress = [
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/securityCodes.ts]: addressesWithoutCommonCodePrefix

File: universal-login-commons/test/core/utils/securityCodes.ts [hunk 7]
```diff
@@ -36,6 +36,70 @@ describe('UNIT: security codes', () => {
...
+      '0x9a2c510AA7E56B83AFe6834f83C24512bafD7318',
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/securityCodes.ts]: '0x9a2c510AA7E56B83AFe6834f83C24512bafD7318', // CODE: [1,2,3,..]

File: universal-login-commons/test/core/utils/securityCodes.ts [hunk 8]
```diff
@@ -36,6 +36,70 @@ describe('UNIT: security codes', () => {
...
+      '0x69a4de9CF98FFf3D019F85C2d0aeb3860C179EBC'
+    ];
+
+    const attackerAddresses = addressesThatDontShareBytesWithRealAddress.concat(addressesThatShareFirstCodeWithRealAddress);
+    const incomingAddresses = attackerAddresses.concat([realAddress]);
+
+    describe('findPossibleAddressesFromPartCode', () => {
+      it('addresses found after 1 code entered', () => {
+        const codeWithOneElement = expectedCode.slice(0, 1);
+        const expectedPossibleAddresses = addressesThatShareFirstCodeWithRealAddress.concat([realAddress]);
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/securityCodes.ts]: [...]

File: universal-login-commons/lib/core/utils/arrays.ts [hunk 9]
```diff
@@ -21,3 +23,19 @@ export const array8bitTo16bit = (numbers: number[]) => {
...
+export const deepArrayStartWith = (array: any[], prefix: any[]) => {
+  if (prefix.length > array.length) {
+    return false;
+  }
+  const arrayPrefix = array.slice(0, prefix.length);
+  return deepEqual(arrayPrefix, prefix);
+};
+
+export const getArrayElementsFromIndicies = <T>(array: T[], indicies: number[]) =>
+  indicies.reduce((elements, index) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/arrays.ts]: `return indicies.map((index) => array[index])`?

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 10]
```diff
@@ -32,6 +29,31 @@ export const isValidCode = (code: number[], publicKey: string) => {
...
+};
+
+export const findPossibleAddressesFromCodePart = (enteredCode: number[], incomingPublicKeys: string[]) => {
+  const generatedCodes = incomingPublicKeys.map((key) => generateCode(key));
+  const validIndicies = findIndiciesOfValidCodesByPrefix(generatedCodes, enteredCode);
+  return getArrayElementsFromIndicies(incomingPublicKeys, validIndicies);
+};
+
+export const findIndiciesOfValidCodesByPrefix = (codes: number[][], enteredCode: number[]) =>
+  codes.reduce((indices, code, index) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]: If you want to do reduce on arrays, you probably should do map and filter instead

Comment (marekkirejczyk) [reply universal-login-commons/lib/core/utils/securityCodes.ts]: Did you mean?
```ts
functoin filterAddressesByPrefix(prefix: [], addresses : [][]) {
  addresses
    .map((address, index) => [address, index])
    .filter(([address, index]) => startsWith(prefix, address))
    .map(([, index]) => index);
}
```

Comment (gasiortomasz) [reply universal-login-commons/lib/core/utils/securityCodes.ts]: Yeah, I meant that ;)

File: universal-login-commons/lib/core/utils/arrays.ts [hunk 11]
```diff
@@ -21,3 +23,16 @@ export const array8bitTo16bit = (numbers: number[]) => {
...
+
+export const deepArrayStartWith = (array: any[], prefix: any[]) => {
+  if (prefix.length > array.length) {
+    return false;
+  }
+  const arrayPrefix = array.slice(0, prefix.length);
+  return deepEqual(arrayPrefix, prefix);
+};
+
+export const getArrayElementsFromIndicies = <T>(array: T[], indicies: number[]) =>
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/arrays.ts]: `indicies` -> `indices`

File: universal-login-commons/lib/core/utils/arrays.ts [hunk 12]
```diff
@@ -21,3 +23,16 @@ export const array8bitTo16bit = (numbers: number[]) => {
...
+  indicies
+    .filter((index) => 0 <= index && index < array.length)
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/arrays.ts]: Fail fast? 
```ts
for (let index of indices) {
  ensure(0 <= index && index < array.length, OutOfBound)
}
```

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 13]
```diff
@@ -32,6 +29,28 @@ export const isValidCode = (code: number[], publicKey: string) => {
...
 
+export const findValidAddressFromCodePart = (enteredCode: number[], incomingPublicKeys: string[]) => {
+  const possibleAddresses = findPossibleAddressesFromCodePart(enteredCode, incomingPublicKeys);
+  return possibleAddresses.length === 1 ? possibleAddresses[0] : undefined;
+};
+
+export const findPossibleAddressesFromCodePart = (enteredCode: number[], incomingPublicKeys: string[]) => {
+  const generatedCodes = incomingPublicKeys.map((key) => generateCode(key));
+  const validIndicies = findIndiciesOfValidCodesByPrefix(generatedCodes, enteredCode);
+  return getArrayElementsFromIndicies(incomingPublicKeys, validIndicies);
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]: Why do we need to involve indices? Why not simply:
```ts
export const findPossibleAddressesFromCodePart = (prefix: number[], publicKeys: string[]) => {
  const codes = publicKeys.map((key) => generateCode(key));
  return filterAddressesByPrefix(codes, prefix);
}
```

That should save a lot of lines of code.

Comment (gasiortomasz) [reply universal-login-commons/lib/core/utils/securityCodes.ts]: Client needs an address so he can confirm the request. If you return a code then client would have to go through all notifications to find which address matches up with that returned code. It is ugly to force this on client, no? It is also extra code and more convoluted step, IMHO.

File: universal-login-commons/lib/core/utils/arrays.ts [hunk 14]
```diff
@@ -21,3 +24,16 @@ export const array8bitTo16bit = (numbers: number[]) => {
...
+
+export const deepArrayStartWith = (array: any[], prefix: any[]) => {
+  if (prefix.length > array.length) {
+    return false;
+  }
+  const arrayPrefix = array.slice(0, prefix.length);
+  return deepEqual(arrayPrefix, prefix);
+};
+
+export const getArrayElementsFromIndicies = <T>(array: T[], indices: number[]) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/arrays.ts]: Indicies -> Indices

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 15]
```diff
@@ -32,6 +36,23 @@ export const isValidCode = (code: number[], publicKey: string) => {
   return deepEqual(expectedCode, code);
 };
 
+export const filterAddressesByCodePrefix = (notifications: Notification[], codePrefix: number[]) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]: filterNotificationsByCodePrefix

**Pull request merged**